basetest.pm

```diff
@@ -640,25 +640,26 @@ sub parse_serial_output_qemu {
     my $json = myjsonrpc::read_json($autotest::isotovideo);
 
     my $die = 0;
+    my %regexp_matched;
     # loop line by line
     for my $line (split(/^/, $json->{serial})) {
         chomp $line;
-        # only two keys allowed
-        for my $type (qw(soft hard)) {
-            for my $regexp (@{$failures->{$type}}) {
-                # If you want to match a simple string please be sure that you create it with quotemeta
-                if ($line =~ /$regexp/) {
-                    my $fail_type = 'softfail';
-                    if ($type eq 'hard') {
-                        $die       = 1;
-                        $fail_type = 'fail';
-                    }
-
-                    $self->record_testresult($fail_type);
-                    $self->record_resultfile('Serial Failure', "Serial error: $line", result => $fail_type);
-                    $self->{result} = $fail_type;
-
+        for my $regexp_table (@{$failures}) {
+            # If you want to match a simple string please be sure that you create it with quotemeta
+            my $regexp = $regexp_table->{pattern};
+            my $type   = $regexp_table->{type};
+            die "Wrong type defined for serial failure. Only soft or hard allowed." if $type !~ /soft|hard/;
+            if (!exists $regexp_matched{$regexp} and $line =~ /$regexp/) {
+                $regexp_matched{$regexp} = 1;
+                my $fail_type = 'softfail';
+                if ($type eq 'hard') {
+                    $die       = 1;
+                    $fail_type = 'fail';
                 }
+
+                $self->record_testresult($fail_type);
+                $self->record_resultfile('Serial Failure', $regexp_table->{message} . " - Serial error: $line", result => $fail_type);
```

Comment:
We need to check if message key is defined, as if not, we'll get an error as try to concatenate uninitialized value.